compiler/rustc_mir_transform/src/fold.rs

```diff
@@ -0,0 +1,357 @@
+            continue;
+        }
+
+        /*
+        if let Rvalue::Cast(CastKind::PtrToPtr, temp_op, _) = temp_rvalue {
+            if let Rvalue::Cast(CastKind::PtrToPtr, _, final_ty) = final_rvalue {
+                let new = Rvalue::Cast(CastKind::PtrToPtr, temp_op.clone(), *final_ty);
+                let new = StatementKind::Assign(Box::new((*final_place, new)));
+
+                let final_local = final_place.local;
+
+                let statements = &mut body.basic_blocks_mut()[temp_loc.block].statements;
+                statements[temp_loc.statement_index].kind = new;
+                statements[final_loc.statement_index].make_nop();
+
+                expand_live_range(body, final_loc.block, final_local);
+
+                return true;
+            }
+        }
+        */
+    }
+    // If we have invalidated anything, we did an optimization
+    !invalidated_statements.is_empty()
+}
+
+fn expand_live_range<'tcx>(body: &mut Body<'tcx>, block: BasicBlock, r#final: Local) {
+    use StatementKind::{StorageDead, StorageLive};
+
+    let mut deleted_live = false;
+    let mut deleted_dead = false;
+
+    for statement in &mut body.basic_blocks_mut()[block].statements {
+        if statement.kind == StorageLive(r#final) {
+            statement.make_nop();
+            deleted_live = true;
+        } else if statement.kind == StorageDead(r#final) {
+            statement.make_nop();
+            deleted_dead = true;
+        }
+    }
+
+    let source_info = body.local_decls[r#final].source_info;
+    let block = &mut body.basic_blocks_mut()[block].statements;
+    if deleted_live {
+        block.insert(0, Statement { source_info, kind: StorageLive(r#final) });
+    }
+    if deleted_dead {
+        block.push(Statement { source_info, kind: StorageDead(r#final) });
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Status {
+    None,
+    Once(Location),
+    Multiple,
+}
+
+impl Status {
+    fn incr(&mut self, location: Location) {
+        *self = match self {
+            Status::None => Self::Once(location),
+            Status::Once(_) | Status::Multiple => Self::Multiple,
+        };
+    }
+
+    fn as_assign<'a, 'tcx>(
+        self,
+        body: &'a Body<'tcx>,
+    ) -> Option<(Location, &'a (Place<'tcx>, Rvalue<'tcx>))> {
+        let location = match self {
+            Status::Once(location) => location,
+            Status::None | Status::Multiple => return None,
+        };
+
+        body.stmt_at(location).left()?.kind.as_assign().map(|res| (location, res))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Analysis {
+    read: Status,
+    write: Status,
+}
+
+impl Default for Analysis {
+    fn default() -> Self {
+        Self { read: Status::None, write: Status::None }
+    }
+}
+
+struct AnalysisVisitor {
+    analysis: IndexVec<Local, Analysis>,
+}
+
+impl<'tcx> Visitor<'tcx> for AnalysisVisitor {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
```

Comment:
You may need to also use a `visit_local` to catch reads in `PlaceElem::Index`, and `TerminatorKind::Return`.

Reply:
Return terminators can have a read?

Reply:
Yes, they read `_0`.